include/sundials/sundials_errors.h

```diff
@@ -0,0 +1,187 @@
+#include <sundials/sundials_types.h>
+
+/* ----------------------------------------------------------------------------
+ * Error code definitions
+ * ---------------------------------------------------------------------------*/
+
+/* SUN_ERR_CODE_LIST is an X macro that can be expanded in various ways */
+#define SUN_ERR_CODE_LIST(ENTRY)                                               \
+  ENTRY(SUN_ERR_ARG_CORRUPT, "argument provided is NULL or corrupted")         \
+  ENTRY(SUN_ERR_ARG_INCOMPATIBLE, "argument provided is not comptaible")       \
+  ENTRY(SUN_ERR_ARG_OUTOFRANGE, "argument is out of the valid range")          \
+  ENTRY(SUN_ERR_ARG_WRONGTYPE, "argument provided is not the right type")      \
+  ENTRY(SUN_ERR_ARG_DIMSMISMATCH, "argument dimensions do not agree")          \
+  ENTRY(SUN_ERR_LOGGER_CORRUPT, "SUNLogger is NULL or corrupt")                \
+  ENTRY(SUN_ERR_LOGGER_CANNOTOPENFILE,                                         \
+        "File provided to SUNLogger could not be opened")                      \
+  ENTRY(SUN_ERR_MALLOC_FAIL, "malloc returned NULL")                           \
+  ENTRY(SUN_ERR_MANYVECTOR_COMMNOTSAME,                                        \
+        "not all subvectors have the same MPI_Comm")                           \
+  ENTRY(SUN_ERR_MANYVECTOR_COMMNULL, "MPI_Comm is NULL")                       \
+  ENTRY(SUN_ERR_MPI_FAIL,                                                      \
+        "the MPI call returned something other than MPI_SUCCESS")              \
+  ENTRY(SUN_ERR_NOT_IMPLEMENTED,                                               \
+        "operation is not implemented: function pointer is NULL")              \
+  ENTRY(SUN_ERR_PROFILER_MAPFULL,                                              \
+        "too many SUNProfiler entries, try setting SUNPROFILER_MAX_ENTRIES "   \
+        "in environment to a bigger number")                                   \
+  ENTRY(SUN_ERR_PROFILER_MAPGET, "unknown error getting SUNProfiler timer")    \
+  ENTRY(SUN_ERR_PROFILER_MAPINSERT,                                            \
+        "unknown error inserting SUNProfiler timer")                           \
+  ENTRY(SUN_ERR_PROFILER_MAPKEYNOTFOUND, "timer was not found in SUNProfiler") \
+  ENTRY(SUN_ERR_PROFILER_MAPSORT, "error sorting SUNProfiler map")             \
+  ENTRY(SUN_ERR_SUNCTX_CORRUPT, "SUNContext is NULL or corrupt")               \
+  ENTRY(SUN_ERR_GENERIC, "")                                                   \
+  ENTRY(SUN_ERR_UNKNOWN, "Unknown error occured: open an issue at "            \
+                         "https://github.com/LLNL/sundials")
+
+/* Expand SUN_ERR_CODE_LIST to enum */
+#define SUN_EXPAND_TO_ENUM(name, description) name,
+
+/* SUNErrorCode range is [-1000, -2000] to avoid conflicts with package error
+   codes, and old/deprecated codes for matrix and (non)linear solvers. */
+enum
+{
+  SUN_ERR_MINIMUM                                       = -2000,
```

Comment:
Strictly speaking, no. However, it helps formalize the error code range beyond us just stating that its [-1000, 2000].

Reply:
My concern with setting the minimum to a specific value is updating this if necessary could easily be forgotten. Is there a way to have `SUN_ERR_MINUMUM` be the same as the last error in the list?

Reply:
Actually, I was mistaken in my previous comment. The minimum is needed to define where the enum values should begin enumerating from. What if we just make the error range larger? Its unlikely we will ever have 1000 different error codes already, but we can make the range [-10000, -1000] and then we really should never need to update the minimum.

Reply:
I was mistaken as well, I was thinking of the values as starting at maximum and going down rather than minimum and going up. So `SUN_ERR_MAXIMUM` should never change but `SUN_ERR_MINIMUM` could need to change if more values are ever needed. I think something like `[-10000, -1000]` sounds good. I could imagine there being a large number of error values if these are ever expanded to include the package return values as well.

Reply:
If we want the minimum to always be one less than the smallest error code value (and similar for max value) then we need to discard the X-macro style and just define an error code and corresponding error code message separately.

Reply:
The x-macro approach seem too useful to abandon here. Maybe having a test to make sure we do not accidentally exceed the range of values would be sufficient.